File: project/app/src/main/java/org/owntracks/android/model/FusedContact.java
package org.owntracks.android.model;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.BaseObservable;
import androidx.databinding.Bindable;
import androidx.lifecycle.MutableLiveData;

import com.google.android.gms.maps.model.LatLng;

import org.owntracks.android.BR;
import org.owntracks.android.geocoding.GeocoderProvider;
import org.owntracks.android.model.messages.MessageCard;
import org.owntracks.android.model.messages.MessageLocation;

import timber.log.Timber;

public class FusedContact extends BaseObservable implements Comparable<FusedContact> {
    private final String id;
    private GeocoderProvider geocoderProvider;
    private MutableLiveData<MessageLocation> messageLocation = new MutableLiveData<>();
    private MessageCard messageCard;
    private Integer imageProvider = 0;
    private long tst = 0;

    @Bindable
    public Integer getImageProvider() {
        return imageProvider;
    }


    @Bindable
    public void setImageProvider(Integer imageProvider) {
        this.imageProvider = imageProvider;
    }

    public FusedContact(@Nullable String id) {
        this.id = (id != null && !id.isEmpty()) ? id : "NOID";
        this.geocoderProvider = geocoderProvider;
    }

    public boolean setMessageLocation(MessageLocation messageLocation) {
        if (tst > messageLocation.getTimestamp())
            return false;

        Timber.v("update contact:%s, tst:%s", id, messageLocation.getTimestamp());

        messageLocation.setContact(this); // Allows to update fusedLocation if geocoder of messageLocation changed
        this.messageLocation.postValue(messageLocation);
        this.tst = messageLocation.getTimestamp();
        notifyMessageLocationPropertyChanged();
        return true;
    }

    public void setMessageCard(MessageCard messageCard) {
        this.messageCard = messageCard;
        notifyMessageCardPropertyChanged();
    }

    private void notifyMessageCardPropertyChanged() {
        this.notifyPropertyChanged(BR.fusedName);
        this.notifyPropertyChanged(BR.imageProvider);
        this.notifyPropertyChanged(BR.id);

    }

    public void notifyMessageLocationPropertyChanged() {
        if (this.messageLocation.getValue() != null) {
            Timber.d("Geocode location updated for %s: %s", this.id, this.messageLocation.getValue().getGeocode());
        }
        this.notifyPropertyChanged(BR.fusedName);
        this.notifyPropertyChanged(BR.messageLocation);
        this.notifyPropertyChanged(BR.geocodedLocation);
        this.notifyPropertyChanged(BR.fusedLocationAccuracy);
        this.notifyPropertyChanged(BR.tst);
        this.notifyPropertyChanged(BR.trackerId);
        this.notifyPropertyChanged(BR.id);
    }


    @Bindable
    public MessageCard getMessageCard() {
        return messageCard;
    }

    @Bindable
    public MutableLiveData<MessageLocation> getMessageLocation() {
        return messageLocation;
    }

    @Bindable
    public String getFusedName() {
        if (hasCard() && getMessageCard().hasName())
            return getMessageCard().getName();
        else
            return getTrackerId();
    }

    @Bindable
    public String getFusedLocationAccuracy() {
        return Integer.toString(this.hasLocation() ? messageLocation.getValue().getAccuracy() : 0);
    }

    @Bindable
    public String getGeocodedLocation() {
        return this.messageLocation.getValue().getGeocode();
    }

    public boolean hasLocation() {
        return this.messageLocation.getValue() != null;
    }

    public boolean hasCard() {
        return this.messageCard != null;
    }


    @Bindable
    @NonNull
    public String getTrackerId() {
        if (hasLocation() && getMessageLocation().getValue().hasTrackerId())
            return getMessageLocation().getValue().getTrackerId();
        else {
            String id = getId().replace("/", "");
            if (id.length() > 2) {
                return id.substring(id.length() - 2);
            } else
                return id;
        }
    }


    @Bindable
    public @NonNull
    String getId() {
        return id;
    }

    public LatLng getLatLng() {
        return new LatLng(this.messageLocation.getValue().getLatitude(), this.messageLocation.getValue().getLongitude());
    }

    private boolean deleted;

    public boolean isDeleted() {
        return deleted;
    }

    public void setDeleted() {
        this.deleted = true;
    }

    @Bindable
    public long getTst() {
        return tst;
    }

    @Override
    public int compareTo(@NonNull FusedContact o) {
        return Long.compare(o.tst, this.tst);
    }
}

File: project/app/src/main/java/org/owntracks/android/model/FusedContact.kt
package org.owntracks.android.model

import androidx.databinding.BaseObservable
import androidx.databinding.Bindable
import org.owntracks.android.BR
import org.owntracks.android.location.LatLng
import org.owntracks.android.model.messages.MessageCard
import org.owntracks.android.model.messages.MessageLocation
import timber.log.Timber

class FusedContact(id: String?) : BaseObservable() {
    @get:Bindable
    val id: String = if (!id.isNullOrEmpty()) id else "NOID"

    @get:Bindable
    var messageLocation: MessageLocation? = null
    internal var messageCard: MessageCard? = null

    @get:Bindable
    var tst: Long = 0
        private set

    fun setMessageLocation(messageLocation: MessageLocation): Boolean {
        if (tst > messageLocation.timestamp) return false
        Timber.v("update contact:%s, tst:%s", id, messageLocation.timestamp)
        messageLocation.setContact(this) // Allows to update fusedLocation if geocoder of messageLocation changed
        this.messageLocation = messageLocation
        tst = messageLocation.timestamp
        notifyMessageLocationPropertyChanged()
        return true
    }

    fun notifyMessageLocationPropertyChanged() {
        if (messageLocation != null) {
            Timber.d("Geocode location updated for %s: %s", id, messageLocation!!.geocode)
        }
        notifyPropertyChanged(BR.fusedName)
        notifyPropertyChanged(BR.messageLocation)
        notifyPropertyChanged(BR.geocodedLocation)
        notifyPropertyChanged(BR.fusedLocationAccuracy)
        notifyPropertyChanged(BR.tst)
        notifyPropertyChanged(BR.trackerId)
        notifyPropertyChanged(BR.id)
    }

    @get:Bindable
    val geocodedLocation: String?
        get() = messageLocation?.geocode

    @get:Bindable
    val fusedName: String
        get() = (messageCard?.name ?: trackerId)

    @get:Bindable
    val fusedLocationAccuracy: Int
        get() = messageLocation?.accuracy ?: 0

    @get:Bindable
    val trackerId: String
        get() = messageLocation?.trackerId ?: id.replace("/", "")
            .let {
                return if (it.length > 2) {
                    it.substring(it.length - 2)
                } else {
                    it
                }
            }

    val latLng: LatLng?
        get() = messageLocation?.run { LatLng(latitude, longitude) }

    @get:Bindable
    val latLngString: String?
        get() = this.latLng?.toString()

    override fun toString(): String {
        return "FusedContact $id ($fusedName)"
    }
}
